enclave/core/sgx/calls.c

```diff
@@ -135,6 +136,160 @@ extern bool oe_disable_debug_malloc_check;
+
+    /* Make sure data is outside of enclave to prevent side-channel attacks
+     * and copy data to local buffers so the host can't change its mind. */
+    local_enclave_plus = enclave_plus;
+    local_init_data = local_enclave_plus->initialization_data;
+    local_config_data = local_init_data ? local_init_data->data : NULL;
+    local_config_size = local_init_data ? local_init_data->size : 0;
+
+    if (!oe_is_outside_enclave(
+            local_enclave_plus,
+            sizeof(oe_enclave_with_initialization_data_t)) ||
+        (local_init_data &&
+         (!oe_is_outside_enclave(
+              local_init_data, sizeof(oe_enclave_initialization_data_t)) ||
+          !oe_is_outside_enclave(local_config_data, local_config_size))))
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(
+        _copy_initialization_data(local_config_data, local_config_size, &data));
+    have_data = data.data && data.size;
+    have_empty_data = local_init_data && !have_data;
+
+    /* Without any initialization data, we assume there is no EEID at all.
+     * We can check the base image property, but accessing the EEID page
+     * would trigger a segfault. */
+    if (!have_data && !have_empty_data)
+        return have_base_image_props ? OE_VERIFY_FAILED : OE_OK;
+
+    /* From here on, we assume EEID is present. If it is missing,
+     * oe_get_eeid() will trigger a segfault. */
+
+    /* Check that the EEID (meta-)data is within the enclave. */
+    eeid = (oe_eeid_t*)oe_get_eeid();
+    if (!oe_is_within_enclave(eeid, OE_PAGE_SIZE))
+        oe_abort();
+
+    /* Check that the signature is within the EEID page. */
+    if (sizeof(oe_eeid_t) + eeid->signature_size > OE_PAGE_SIZE)
+        oe_abort();
+
+    /* Check that the EEID version matches. */
+    if (eeid->version != OE_EEID_VERSION)
+        oe_abort();
+
+    /* Check that the entry-point is as expected. */
+    if (eeid->entry_point != oe_enclave_properties_sgx.image_info.entry_rva)
+        oe_abort();
+
+    if (have_data)
+    {
+        /* Check that the data hashes to the correct value. */
+        OE_SHA256 config_hash;
+        oe_sha256(data.data, data.size, &config_hash);
+        if (memcmp(config_hash.buf, eeid->config_id, OE_SHA256_SIZE) != 0)
+            OE_RAISE(OE_VERIFY_FAILED);
+
+        /* Copy the initialization data into enclave heap. */
+        _enclave_initialization_data =
+            oe_malloc(sizeof(oe_enclave_initialization_data_t));
+        if (!_enclave_initialization_data)
+            OE_RAISE(OE_OUT_OF_MEMORY);
+        _enclave_initialization_data->data = data.data;
+        _enclave_initialization_data->size = data.size;
+        data.data = NULL;
+        data.size = 0;
+    }
+    else
+    {
+        /* We have EEID with a config_id, but no data. It is up to the user to
+         * check that the config_id is as expected. */
+    }
+
+    eeid_checked = true;
+    result = OE_OK;
+
+done:
+    oe_free(data.data);
```

Comment:
In several error paths, `data.data` is NULL. It's also explictly set to NULL on line 263.

Reply:
That's right and there's no problem with that. `free` or `oe_free` handle `NULL` as expected.

Reply:
OK